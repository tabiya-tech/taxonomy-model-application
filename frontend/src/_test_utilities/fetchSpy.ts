export function setupFetchSpy(
  expectedStatus: number,
  expectedResponseBody: string | object | undefined,
  contentType: "" | "application/json;charset=UTF-8"
): jest.SpyInstance {
  const responseBody =
    typeof expectedResponseBody === "string" ? expectedResponseBody : JSON.stringify(expectedResponseBody);
  const expectedResponse = new Response(responseBody, {
    status: expectedStatus,
    headers: { "Content-Type": contentType },
  });
  return jest.spyOn(window, "fetch").mockResolvedValue(expectedResponse);
}
