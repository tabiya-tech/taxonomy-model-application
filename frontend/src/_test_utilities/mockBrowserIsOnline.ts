import { fireEvent } from "@testing-library/react";

export function mockBrowserIsOnLine(isOnline: boolean) {
  if (window.navigator.onLine !== isOnline) {
    jest.spyOn(window.navigator, "onLine", "get").mockReturnValue(isOnline);
    if (isOnline) {
      fireEvent.online(window);
    } else {
      fireEvent.offline(window);
    }
  }
}

export function unmockBrowserIsOnLine() {
  jest.spyOn(window.navigator, "onLine", "get").mockRestore();
}
