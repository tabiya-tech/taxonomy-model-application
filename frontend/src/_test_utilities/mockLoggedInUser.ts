import { AuthContextValue, TabiyaUser, TabiyaUserRole, authContextDefaultValue } from "src/auth/AuthProvider";

export function mockLoggedInUser(mockedValues: Partial<AuthContextValue>){
  let user: TabiyaUser = mockedValues.user || { username: "foo", roles: [] };
  jest.spyOn(require("src/auth/hooks/useAuthUser"), "useAuthUser").mockReturnValue({
    ...authContextDefaultValue,
    user: user,
    setUser: (user: TabiyaUser) => null,
    hasRole: (role: TabiyaUserRole) => user.roles.includes(role),
    updateUserByAccessToken: () => {},
    ...mockedValues
  });
}
