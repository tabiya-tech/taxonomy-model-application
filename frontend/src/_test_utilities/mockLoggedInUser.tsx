import {
  AuthContextValue,
  TabiyaUser,
  authContextDefaultValue,
  AuthContext,
  AuthProviderProps,
} from "src/auth/AuthProvider";
import AuthAPISpecs from "api-specifications/auth";

export function mockLoggedInUser(mockedValues: Partial<AuthContextValue>) {
  let user: TabiyaUser = mockedValues.user || TestUsers.Anonymous;

  const value = {
    ...authContextDefaultValue,
    user: user,
    setUser: (user: TabiyaUser) => null,
    hasRole: (role: AuthAPISpecs.Enums.TabiyaRoles) => user.roles.includes(role),
    updateUserByAccessToken: () => {},
    ...mockedValues,
  };

  // @ts-ignore
  jest.spyOn(require("src/auth/AuthProvider"), "AuthProvider").mockImplementation((props: AuthProviderProps) => {
    return <AuthContext.Provider value={value}>{props.children}</AuthContext.Provider>;
  });
}

export const TestUsers = {
  ModelManager: {
    username: "ModelManagerUser",
    roles: [AuthAPISpecs.Enums.TabiyaRoles.MODEL_MANAGER],
  },
  RegisteredUser: {
    username: "RegisteredUser",
    roles: [AuthAPISpecs.Enums.TabiyaRoles.REGISTERED_USER],
  },
  Anonymous: {
    username: "AnonymousUser",
    roles: [AuthAPISpecs.Enums.TabiyaRoles.ANONYMOUS],
  },
};
