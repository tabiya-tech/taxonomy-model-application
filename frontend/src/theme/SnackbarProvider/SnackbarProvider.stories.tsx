import type { Meta, StoryObj } from "@storybook/react";
import SnackbarProvider, { useSnackbar, VariantType } from "./SnackbarProvider";
import { Button } from "@mui/material";
import { v4 as uuidv4 } from "uuid";
import { faker } from "@faker-js/faker";
import { useEffect } from "react";

const meta: Meta<typeof SnackbarProvider> = {
  title: "Components/SnackBar",
  component: SnackbarProvider,
  tags: ["autodocs"],
  argTypes: {},
  parameters: {
    docs: { disable: true },
    a11y: {
      // Disabling a11y due to https://github.com/iamhosseindhv/notistack/issues/579
      // See also frontend/src/theme/SnackbarProvider/SnackbarProvider.tsx
      disable: true,
    },
  },
};

export default meta;

type Story = StoryObj<typeof SnackbarProvider>;

type Color = "success" | "error" | "warning" | "info";

interface ButtonConfig {
  color: Color;
  message: string;
  variant: VariantType;
}

const TestButton = () => {
  const { enqueueSnackbar } = useSnackbar();

  const buttons: ButtonConfig[] = [
    { color: "success", message: "Success message", variant: "success" },
    { color: "error", message: "Failed message", variant: "error" },
    { color: "warning", message: "Warning message", variant: "warning" },
    { color: "info", message: "Info message", variant: "info" },
  ];

  const handleSnackbar = (message: string, variant: VariantType) => {
    enqueueSnackbar(message, { variant });
  };

  return (
    <>
      {buttons.map((button: ButtonConfig) => (
        <Button
          key={uuidv4()}
          color={button.color}
          onClick={() => handleSnackbar(button.message, button.variant)}
          style={{ textTransform: "uppercase" }}
        >
          {button.variant}
        </Button>
      ))}
    </>
  );
};
export const Shown: Story = {
  // exclude from a11y tests as this is for visual demonstration only
  parameters: {
    docs: { disable: false },
    a11y: { disable: true },
  },
  render: () => <TestButton />,
};

export const ShownInfo: Story = {
  render: () => {
    return <ShowNotifications notifications={getRandomNotificationProps(5, "info")} />;
  },
};

export const ShownWarning: Story = {
  render: () => {
    return <ShowNotifications notifications={getRandomNotificationProps(5, "warning")} />;
  },
};

export const ShownError: Story = {
  render: () => {
    return <ShowNotifications notifications={getRandomNotificationProps(5, "error")} />;
  },
};

export const ShownSuccess: Story = {
  render: () => {
    return <ShowNotifications notifications={getRandomNotificationProps(5, "success")} />;
  },
};

export const ShownMixed: Story = {
  render: () => {
    const notifications = [
      ...getRandomNotificationProps(3, "info"),
      ...getRandomNotificationProps(3, "warning"),
      ...getRandomNotificationProps(3, "error"),
      ...getRandomNotificationProps(3, "success"),
    ];
    return <ShowNotifications notifications={notifications} />;
  },
};

function getRandomNotificationProps(count: number, variant: VariantType): { message: string; variant: VariantType }[] {
  return Array(count)
    .fill(undefined)
    .map(() => ({ message: faker.lorem.words(10), variant: variant }));
}

function ShowNotifications(props: Readonly<{ notifications: { message: string; variant: VariantType }[] }>) {
  const { enqueueSnackbar } = useSnackbar();
  useEffect(() => {
    props.notifications.forEach((notification) => {
      enqueueSnackbar(notification.message, {
        variant: notification.variant,
        transitionDuration: { enter: 0, exit: 0 },
      }); // disable the transition to speed up the test
    });
  });
  return <></>;
}
