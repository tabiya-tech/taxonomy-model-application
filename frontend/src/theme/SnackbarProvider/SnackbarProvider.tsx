import {
  SnackbarProvider as OriginalSnackbarProvider,
  SnackbarProviderProps as OriginalSnackbarProviderProps,
  useSnackbar,
  SnackbarKey,
  MaterialDesignContent,
} from "notistack";
import { IconButton, styled } from "@mui/material";
import CloseIcon from "@mui/icons-material/Close";
import * as React from "react";

const uniqueId = "bade4e92-baed-40f8-a098-0e745563e786";
export const DATA_TEST_ID = {
  SNACKBAR_CLOSE_BUTTON: `snackbar-close-button-${uniqueId}`,
};

const SnackbarCloseButton = (key: SnackbarKey) => {
  const { closeSnackbar } = useSnackbar();
  // Currently due to a bug in the notistack library, we cannot ref to the correct ARIA ids (see https://github.com/iamhosseindhv/notistack/issues/579)
  // Suing  the title attribute for accessibility, is not ideal. When multiple snackbars are open,
  // the title will be the same for all of them violating the WAG2A rule of unique id. See https://tabiya-tech.atlassian.net/browse/PLAT-129
  return (
    <IconButton
      title={"Close notification"}
      data-testid={DATA_TEST_ID.SNACKBAR_CLOSE_BUTTON}
      onClick={() => closeSnackbar(key)}
      sx={{
        margin: "0",
        padding: "0",
      }}
    >
      <CloseIcon
        sx={{
          margin: 0,
          padding: 0,
        }}
      />
    </IconButton>
  );
};

const StyledMaterialDesignContent = styled(MaterialDesignContent)(({ theme }) => ({
  "&.notistack-MuiContent": {
    // adjust the snackbar content to place the close button on the top right,
    // so that the button is not hidden in case multiple snackbars are open and overlap,
    // especially when the text is long
    display: "flex",
    flexDirection: "row",
    flexWrap: "nowrap",
    alignItems: "start",
    // ---
  },
  "&.notistack-MuiContent-success": {
    color: theme.palette.success.dark,
    backgroundColor: theme.palette.success.light,
    fontFamily: theme.typography.body1.fontFamily,
    fontSize: theme.typography.body1.fontSize,
    "& .MuiSvgIcon-root": {
      color: theme.palette.success.dark,
    },
  },
  "& #notistack-snackbar svg": {
    fontSize: theme.typography.h4.fontSize + "!important",
  },
  "&.notistack-MuiContent-error": {
    color: theme.palette.error.dark,
    backgroundColor: theme.palette.error.light,
    fontFamily: theme.typography.body1.fontFamily,
    fontSize: theme.typography.body1.fontSize,
    "& .MuiSvgIcon-root": {
      color: theme.palette.error.dark,
    },
  },
  "&.notistack-MuiContent-warning": {
    color: theme.palette.warning.dark,
    backgroundColor: theme.palette.warning.light,
    fontFamily: theme.typography.body1.fontFamily,
    fontSize: theme.typography.body1.fontSize,
    "& .MuiSvgIcon-root": {
      color: theme.palette.warning.dark,
    },
  },
  "&.notistack-MuiContent-info": {
    color: theme.palette.info.dark,
    backgroundColor: theme.palette.info.light,
    fontFamily: theme.typography.body1.fontFamily,
    fontSize: theme.typography.body1.fontSize,
    "& .MuiSvgIcon-root": {
      color: theme.palette.info.dark,
    },
  },
}));

const SnackbarProvider: React.FC<OriginalSnackbarProviderProps> = ({
  children,
  ...props
}: Readonly<OriginalSnackbarProviderProps>) => {
  const notistackOptions: OriginalSnackbarProviderProps = {
    maxSnack: 10,
    autoHideDuration: null,
    transitionDuration: { enter: 500, exit: 500 },
    anchorOrigin: {
      vertical: "top",
      horizontal: "right",
    },
    Components: {
      success: StyledMaterialDesignContent,
      error: StyledMaterialDesignContent,
      warning: StyledMaterialDesignContent,
      info: StyledMaterialDesignContent,
    },
    dense: false,
    style: {
      // limit the width of the snackbar so that long text is wrapped in multiple lines
      maxWidth: "600px",
      whiteSpace: "normal",
      // ---
    },
    ...props,
  };

  return (
    <OriginalSnackbarProvider {...notistackOptions} action={SnackbarCloseButton}>
      {children}
    </OriginalSnackbarProvider>
  );
};

export default SnackbarProvider;
export { useSnackbar };
export type { VariantType } from "notistack";
