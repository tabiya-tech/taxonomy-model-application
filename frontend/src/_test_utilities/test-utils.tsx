// Based on https://testing-library.com/docs/react-testing-library/setup/

import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "@mui/material";
import applicationTheme, { ThemeMode } from "src/theme/applicationTheme/applicationTheme";
import SnackbarProvider from "src/theme/SnackbarProvider/SnackbarProvider";
import { IsOnlineProvider } from "src/app/providers";

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <IsOnlineProvider>
      <ThemeProvider theme={applicationTheme(ThemeMode.LIGHT)}>
        <SnackbarProvider>{children}</SnackbarProvider>
      </ThemeProvider>
    </IsOnlineProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
