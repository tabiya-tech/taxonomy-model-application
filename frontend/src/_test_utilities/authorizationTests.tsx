import { render, screen } from "src/_test_utilities/test-utils";
import AuthAPISpecs from "api-specifications/auth";
import { mockLoggedInUser } from "./mockLoggedInUser";

export const ALL_USERS = [
  AuthAPISpecs.Enums.TabiyaRoles.ANONYMOUS,
  AuthAPISpecs.Enums.TabiyaRoles.REGISTERED_USER,
  AuthAPISpecs.Enums.TabiyaRoles.MODEL_MANAGER,
];

type AuthorizationTestsParams = {
  name: string;
  Component: JSX.Element;
  testIds?: string[];
  roles?: AuthAPISpecs.Enums.TabiyaRoles[];
  tests?: Partial<
    Record<
      AuthAPISpecs.Enums.TabiyaRoles,
      {
        visible?: string[];
        hidden?: string[];
        visibleDisabled?: string[];
      }
    >
  >;
};

export const authorizationTests = {
  defaultName: "Test the rendering of the component based on Roles",
  callback: ({ name, Component, testIds = [], roles = ALL_USERS, tests }: AuthorizationTestsParams) => {
    return () => {
      test.each(roles?.map((role) => [role]))(`${name} renders correctly for user with role %s`, (role) => {
        const commonElements = testIds || [];
        const { visible = [], visibleDisabled = [], hidden = [] } = tests?.[role] || {};

        mockLoggedInUser({ user: { username: "foo", roles: [role] } });

        render(Component);

        commonElements.forEach((testId) => expect(screen.getByTestId(testId)).toBeInTheDocument());

        visible?.forEach((testId) => {
          expect(screen.getByTestId(testId)).toBeVisible();
        });

        hidden?.forEach((testId) => {
          expect(screen.queryByTestId(testId)).not.toBeInTheDocument();
        });

        visibleDisabled?.forEach((testId) => {
          const element = screen.getByTestId(testId);
          expect(element).toBeVisible();
          if (element instanceof HTMLInputElement) {
            expect(element).toBeDisabled();
          }
        });
      });
    };
  },
};
