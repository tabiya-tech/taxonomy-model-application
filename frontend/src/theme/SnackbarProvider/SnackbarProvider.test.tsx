// mute the console
import "src/_test_utilities/consoleMock";

import React from "react";
import { act, render, screen } from "src/_test_utilities/test-utils";
import { DATA_TEST_ID, useSnackbar } from "./SnackbarProvider";
import userEvent from "@testing-library/user-event";
import { Button } from "@mui/material";

describe("SnackbarProvider render tests", () => {
  beforeEach(() => {
    (console.error as jest.Mock).mockClear();
    (console.warn as jest.Mock).mockClear();
  });

  it("renders children correctly", () => {
    render(<div data-testid="test-child">Test Child</div>);

    // THEN expect no errors or warning to have occurred
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();

    const childElement = screen.getByTestId("test-child");
    expect(childElement).toBeInTheDocument();
  });
});

describe("SnackbarProvider render action tests", () => {
  const TestButton = () => {
    const { enqueueSnackbar } = useSnackbar();
    const handleSnackbar = () => {
      enqueueSnackbar("message", { variant: "success" });
    };
    return (
      <Button data-testid="test-button" onClick={handleSnackbar} style={{ textTransform: "uppercase" }}>
        Test
      </Button>
    );
  };

  it("should close the snackbar when the close button is clicked", async () => {
    // GIVEN a component that is using the snackbar
    render(<TestButton />);

    // WHEN the snackbar is shown
    // useFakeTimers to speed up the test as the snackbar is animated, and we don't want to wait for the animation
    jest.useFakeTimers();
    const userEventFakeTimer = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
    const testButton = screen.getByTestId("test-button");
    await userEventFakeTimer.click(testButton);

    // THEN expect no errors or warning to have occurred
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();
    // AND the snackbar to be visible
    act(() => jest.runOnlyPendingTimers());
    const closeButton = screen.getByTestId(DATA_TEST_ID.SNACKBAR_CLOSE_BUTTON);
    expect(closeButton).toBeVisible();
    // AND the button to match the snapshot
    expect(closeButton).toMatchSnapshot();
    // AND the snackbar close button is clicked
    await userEventFakeTimer.click(closeButton);
    // AND the snackbar should be closed
    act(() => jest.runOnlyPendingTimers());
    expect(closeButton).not.toBeVisible();

    jest.useRealTimers();
  });
});
