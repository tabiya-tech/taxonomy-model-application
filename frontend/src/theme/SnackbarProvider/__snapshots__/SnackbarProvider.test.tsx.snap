// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SnackbarProvider render action tests should close the snackbar when the close button is clicked 1`] = `
<button
  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-8am8gs-MuiButtonBase-root-MuiIconButton-root"
  data-testid="snackbar-close-button-bade4e92-baed-40f8-a098-0e745563e786"
  tabindex="0"
  title="Close notification"
  type="button"
>
  <svg
    aria-hidden="true"
    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-12bfw96-MuiSvgIcon-root"
    data-testid="CloseIcon"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
    />
  </svg>
  <span
    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</button>
`;
